### PR TITLE
Feat/pro 5018/ip discovery

### PR DESCRIPTION
**Pull request by patriotyk**

### CODE REVIEW

File: src/config.h [hunk 1]
```diff
@@ -32,9 +32,14 @@ struct GatewayConfig {
 };
 
 struct NetworkConfig {
-  std::string socket_commands_path{"/tmp/sota-commands.socket"};
-  std::string socket_events_path{"/tmp/sota-events.socket"};
-  std::vector<std::string> socket_events{"DownloadComplete", "DownloadFailed"};
+  NetworkConfig() : socket_commands_path("/tmp/sota-commands.socket"), socket_events_path("/tmp/sota-events.socket") {}
+  std::string socket_commands_path;
+  std::string socket_events_path;
+  std::vector<std::string> socket_events = {"DownloadComplete", "DownloadFailed"};
```
Comment (cajun-rat) [inline src/config.h]: Is there a reason to revert these changes, or is this a rebase error?

Comment (patriotyk) [reply src/config.h]: No sorry, I did not noticed that when resolving conflicts for rebase

Comment (cajun-rat) [reply src/config.h]: Heh, no worries

File: tests/CMakeLists.txt [hunk 2]
```diff
@@ -158,6 +160,9 @@ add_aktualizr_test(NAME uptane_key SOURCES uptane_key_test.cc PROJECT_WORKING_DI
 
 add_aktualizr_test(NAME uptane_network SOURCES uptane_network_test.cc PROJECT_WORKING_DIRECTORY)
 
+add_aktualizr_test(NAME discovery_secondary SOURCES ipsecondary_discovery_test.cc PROJECT_WORKING_DIRECTORY)
+
+
```
Comment (cajun-rat) [inline tests/CMakeLists.txt]: +1 for a test, but it fails on my machine :(

~~~
[ RUN      ] ipsecondary_discovery.test_discovery
starting up on 127.0.0.1 port 60443

waiting to receive message
/home/phil/pro/client/aktualizr/tests/ipsecondary_discovery_test.cc:25: Failure
      Expected: secondaries.size()
      Which is: 0
To be equal to: 2
~~~

**Comment by cajun-rat** [comment]
Overall this looks good, and we can change the broadcast address to a suitable subnet at some point

File: src/ipsecondarydiscovery.cc [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    tv.tv_sec = left_seconds;
+    setsockopt(socket_fd, SOL_SOCKET, SO_RCVTIMEO, (const char *)&tv, sizeof(tv));
+  }
+  return secondaries;
+}
+
+void IpSecondaryDiscovery::sendRequest() {
+  int socket_fd = socket(AF_INET, SOCK_DGRAM, 0);
+  if (socket_fd == -1) {
+    LOG_ERROR << "Error of creating socket: " << std::strerror(errno);
```
Comment (lbonn) [inline src/ipsecondarydiscovery.cc]: The function should return or throw in case of errors (also in subsequent `setsockopt`).

To make that easier, I was about to move `SocketHandle` to `utils.h`:
https://github.com/advancedtelematic/aktualizr/blob/05eeb37d2bfaa0e31cc26a4a6c4d610efd7f8ba9/src/aktualizr_secondary/aktualizr_secondary.h#L17
But no need to use it right away.

File: src/ipsecondarydiscovery.cc [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  tv.tv_usec = 0;
+  setsockopt(socket_fd, SOL_SOCKET, SO_RCVTIMEO, (const char *)&tv, sizeof(tv));
+
+  char rbuf[2000] = {};
+  int recieved = 0;
+  while ((recieved = recv(socket_fd, rbuf, sizeof(rbuf) - 1, 0)) != -1) {
+    std::string data(std::move(rbuf), recieved);
+    try {
+      int type = 0;
+      Uptane::SecondaryConfig conf;
```
Comment (lbonn) [inline src/ipsecondarydiscovery.cc]: The secondary config should probably contain the device ip so that we can communicate with it on the regular interface later.

File: src/ipsecondarydiscovery.cc [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+      asn1::Deserializer des(data);
+      des >> asn1::seq >> asn1::implicit<kAsn1Integer>(type) >> asn1::implicit<kAsn1Utf8String>(conf.ecu_serial) >>
+          asn1::implicit<kAsn1Utf8String>(conf.ecu_hardware_id) >> asn1::restseq;
```
Comment (lbonn) [inline src/ipsecondarydiscovery.cc]: @OYTIS do you think it would be worth it to also transmit the secondary server tcp port, to make the discovery more fail-safe? So that we only need to change the port in one place. We can worry about it later but it's pretty straightforward.

Comment (patriotyk) [reply src/ipsecondarydiscovery.cc]: We can't use discovery with tcp because it requires connection to hosts. But we need to find those hosts.

Comment (lbonn) [reply src/ipsecondarydiscovery.cc]: Yes, I meant transmitting the tcp port number of the server (secondary) in the udp discovery response so that it doesn't need to be hard-coded on the primary side.

Comment (patriotyk) [reply src/ipsecondarydiscovery.cc]: Do we need this IP now? It is not a problem to add any fields we need later.

Comment (lbonn) [reply src/ipsecondarydiscovery.cc]: You're right, I think it can wait until the integration

File: src/ipsecondarydiscovery.cc [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    setsockopt(socket_fd, SOL_SOCKET, SO_RCVTIMEO, (const char *)&tv, sizeof(tv));
+  }
+  return secondaries;
+}
+
+void IpSecondaryDiscovery::sendRequest() {
+  int broadcast = 1;
+  if (setsockopt(socket_fd, SOL_SOCKET, SO_BROADCAST, &broadcast, sizeof broadcast) < 0) {
+    LOG_ERROR << "Could not setup socket for broadcast: " << std::strerror(errno);
+    close(socket_fd);
```
Comment (lbonn) [inline src/ipsecondarydiscovery.cc]: Should return here.

And possibly indicate failure in the method's return type so that aktualizr can avoid waiting 10s if it failed to send the request.

Comment (patriotyk) [reply src/ipsecondarydiscovery.cc]: Yes, You are right. Done

File: src/config.h (hunk 1 above)
Comment (cajun-rat) [reply src/config.h]: Can you un-revert these back to the c++11 style, please?

Comment (patriotyk) [reply src/config.h]: Sorry, I marked this change in my mind as resolved. 10 minutes and will update PR
